packages/@headlessui-react/src/components/focus-trap/focus-trap.tsx

```diff
@@ -210,31 +210,68 @@ export let FocusTrap = Object.assign(FocusTrapRoot, {
 
 // ---
 
-function useRestoreFocus({ ownerDocument }: { ownerDocument: Document | null }, enabled: boolean) {
-  let restoreElement = useRef<HTMLElement | null>(null)
+let history: HTMLElement[] = []
+if (typeof window !== 'undefined' && typeof document !== 'undefined') {
+  function handle(e: Event) {
+    if (!(e.target instanceof HTMLElement)) return
+    if (e.target === document.body) return
+    if (history[0] === e.target) return
+
+    history.unshift(e.target)
+
+    // Filter out DOM Nodes that don't exist anymore
+    history = history.filter((x) => x != null && x.isConnected)
+    history.splice(10) // Only keep the 10 most recent items
+  }
 
-  // Capture the currently focused element, before we try to move the focus inside the FocusTrap.
-  useEventListener(
-    ownerDocument?.defaultView,
-    'focusout',
-    (event) => {
-      if (!enabled) return
-      if (restoreElement.current) return
+  window.addEventListener('click', handle, { capture: true })
+  window.addEventListener('mousedown', handle, { capture: true })
+  window.addEventListener('focus', handle, { capture: true })
+
+  document.body.addEventListener('click', handle, { capture: true })
+  document.body.addEventListener('mousedown', handle, { capture: true })
+  document.body.addEventListener('focus', handle, { capture: true })
+}
 
-      restoreElement.current = event.target as HTMLElement
+function useRestoreElement(enabled: boolean = true) {
+  let localHistory = useRef<HTMLElement[]>(history.slice())
+
+  useWatch(
+    ([newEnabled], [oldEnabled]) => {
+      // We are disabling the restore element, so we need to clear it.
+      if (oldEnabled === true && newEnabled === false) {
+        // However, let's schedule it in a microTask, so that we can still read the value in the
+        // places where we are restoring the focus.
+        microTask(() => {
+          localHistory.current.splice(0)
+        })
```

Comment:
Things look generally good. This is the only thing that feels weird to me. Mind talking through this when you get a chance?

Reply:
We might be able to model this in a different way without the scary `microTask`. But my idea was that I don't want to have stale data if we toggle the `enabled` flag. So the `splice(0)` should clear the current `localHistory` that is a scoped snapshot of the "global" `history`. Because if you request the restore element when the `enabled` flag is `false` I want to get `null` instead of an element.

The reason it is in a `microTask` is because the `useEffect`s are executed in order. And I wanted to defer the cleanup, because otherwise that would result in this behaviour:

1. `useRestoreElement`'s `useEffect` would run, and clears the `localHistory` (because we toggled the `enabled` flag).
2. `useRestoreFocus`'s `useEffect` would run, and uses the callback returned from `useRestoreElement` to get the first still connected element. But at this time the `localHistory` was already cleared. 

I also used a function so that we can delay the calculation of the restore element because it could be that _something_ happened and the next-in-line restore element is not connected to the DOM anymore so we want to take the next one. 

So I considered it like "defer", that the `localHistory` is cleared after we called the function.

We could get rid of this block entirely. The `localHistory` should get GC'd when this is not used anymore. When we switch `enabled` from `false` to `true` it will set the `localHistory.current` to the latest snapshot anyway. It's just when requesting the data while the `enabled` flag is `false` that we don't want to get the incorrect element.

Reply:
Okay, that makes sense. Thanks!